lib/fluent/plugin/out_google_cloud.rb

```diff
@@ -665,14 +688,141 @@ def write(chunk)
+    #           },
+    #           "3": {
+    #             "code": 3,
+    #             "message": "Log name contains illegal character :"
+    #           }
+    #         }
+    #       },
+    #       {
+    #         "@type": "type.googleapis.com/google.rpc.DebugInfo",
+    #         "detail": ...
+    #       }
+    #     ]
+    #   }
+    # }
+    #
+    # The root level "code", "message", and "status" simply match the root
+    # cause of the first failed log entry. For example, if we switch the order
+    # of the log entries, then we would get:
+    # {
+    #    "error" : {
+    #       "code" : 400,
+    #       "message" : "Log name contains illegal character :",
+    #       "status" : "INVALID_ARGUMENT",
+    #       "details": ...
+    #    }
+    # }
+    # We will ignore it anyway and look at the details instead which includes
+    # info for all failed log entries.
+    #
+    # In this example, the logEntryErrors we are extracting should be:
+    # {
+    #   "0": {
+    #     "code": 7,
+    #     "message": "User not authorized."
+    #   },
+    #   "1": {
+    #     "code": 3,
+    #     "message": "Log name contains illegal character :"
+    #   },
+    #   "3": {
+    #     "code": 3,
+    #     "message": "Log name contains illegal character :"
+    #   }
+    # }
+    def extract_log_entry_errors(error)
+      error_details = JSON.parse(error.body)['error']['details']
+      partial_errors = error_details.detect do |error_detail|
+        error_detail['@type'] == PARTIAL_ERROR_FIELD
+      end
+      partial_errors['logEntryErrors']
+    rescue JSON::ParserError => e
+      @log.warn 'Failed to extract log entry errors from the error details' \
+                " because of JSON parsing error: #{error.body}.", error: e
+      {}
+    rescue => e
+      @log.warn 'Failed to extract log entry errors from the error details' \
+                " JSON: #{error.body}.", error: e
+      {}
+    end
+
+    # Given the logEntryErrors, constuct a map from errors to a list of indexes
+    # of log entries that failed for this specific error.
+    #
+    # The keys in this map are [error_code, error_message] pairs. The values in
+    # this map are a list of indexes of log entries that failed due to this
+    # error.
+    #
+    # For logEntryErrors as in the example above (in the header comment for
+    # extract_log_entry_errors), the constructed map would be:
+    # {
+    #   [7, 'User not authorized.']: ['0'],
+    #   [3, 'Log name contains illegal character :']: ['1', '3']
+    # }
+    def construct_error_details_map(log_entry_errors)
+      error_details_map = Hash.new { |h, k| h[k] = [] }
+      log_entry_errors.each do |index, error|
+        error_key = [error['code'], error['message']].freeze
+        # TODO(qingling128): Convert indexes to integers.
+        error_details_map[error_key] << index
+      end
+      error_details_map
+    rescue => e
```

Comment:
What exception are we expecting here?

Reply:
I added the warning to be:

```
      @log.warn 'Failed to parse the log entry errors because logEntryErrors ' \
                "is not a hash or it does not contain 'code' and 'message' " \
                'fields.', error: e
```

Reply:
As I said above, the exception you'd get would not be very useful to the user. You could use the same `ensure_hash` method, e.g.:
```ruby
    def construct_error_details_map(log_entry_errors)
      error_details_map = Hash.new { |h, k| h[k] = [] }
      log_entry_errors.each do |index, error|
        error_hash = ensure_hash(error)
        unless error_hash['code'] and error_hash['message'] then
          @log.warn "Entry with index #{index} is missing 'code' or 'message'"
                    ' field.'
          return {} # Can also be 'next' or 'break'...
        end
        error_key = [error_hash['code'], error_hash['message']].freeze
        # TODO(qingling128): Convert indexes to integers.
        error_details_map[error_key] << index
      end
      error_details_map
    end
```

Reply:
Changed.

Reply:
You still have the unqualified `rescue` block...